e2e/benchmarks/local-benchmark/util.js

```diff
@@ -66,3 +66,102 @@ function queryTimerIsEnabled() {
+  if (!isFinite(a) && !isFinite(e)) {
+    return true;
+  }
+  if (isNaN(a) || isNaN(e) || Math.abs(a - e) > epsilon) {
+    return false;
+  }
+  return true;
+}
+
+function expectObjectsPredicate(actual, expected, epsilon, predicate) {
+  let actualKeys = Object.getOwnPropertyNames(actual);
+  let expectedKeys = Object.getOwnPropertyNames(expected);
+  if (actualKeys.length != expectedKeys.length) {
+    throw new Error(`Actual length ${
+        actualKeys.length} not equal Expected length ${expectedKeys.length}`);
+  }
+  for (let i = 0; i < actualKeys.length; i++) {
+    let key = actualKeys[i];
+    let isObject = typeof (actual[key]) === 'object' &&
+        typeof (expected[key]) === 'object';
+    let isArray = tf.util.isTypedArray(actual[key]) &&
+        tf.util.isTypedArray(expected[key]);
+    if (isArray) {
+      expectArraysClose(actual[key], expected[key], epsilon, key);
+    } else if (!isObject && (!predicate(actual[key], expected[key]))) {
+      throw new Error(`Objects differ: actual[${key}] = ${
+          JSON.stringify(actual[key])}, expected[${key}] = ${
+          JSON.stringify(expected[key])}!`);
+    } else if (
+        isObject &&
+        !expectObjectsPredicate(
+            actual[key], expected[key], epsilon, predicate)) {
+      throw new Error(`Objects differ: actual[${key}] = ${
+          JSON.stringify(actual[key])}, expected[${key}] = ${
+          JSON.stringify(expected[key])}!`);
+    }
+  }
+  return true;
+}
+
+function expectObjectsClose(actual, expected, epsilon) {
+  if (epsilon == null) {
+    epsilon = tf.test_util.testEpsilon();
+  }
+  expectObjectsPredicate(
+      actual, expected, epsilon, (a, b) => areClose(a, b, epsilon));
+}
+
+function expectArraysPredicateFuzzy(actual, expected, predicate, errorRate) {
+  if (tf.util.isTypedArray(actual) == false ||
+      tf.util.isTypedArray(expected) == false) {
+    throw new Error(`Actual and Expected are not arrays.`);
+  }
+
+  if (actual.length !== expected.length) {
+    throw new Error(
+        `Arrays have different lengths actual: ${actual.length} vs ` +
+        `expected: ${expected.length}.\n` +
+        `Actual:   ${actual}.\n` +
+        `Expected: ${expected}.`);
+  }
+  let mismatchCount = 0;
+  for (let i = 0; i < expected.length; ++i) {
+    const a = actual[i];
+    const e = expected[i];
+    if (!predicate(a, e)) {
+      mismatchCount++;
+      const maxMismatch = Math.floor(errorRate * expected.length);
+      if (mismatchCount > maxMismatch) {
+        throw new Error(
+            `Arrays data has more than ${maxMismatch} differs from ${
+                expected.length}: actual[${i}] = ${a}, expected[${i}] = ${
+                e}.\n` +
+            `Actual:   ${actual}.\n` +
+            `Expected: ${expected}.`);
+      }
+    }
+  }
+}
+
+function expectArraysClose(actual, expected, epsilon, key) {
+  if (epsilon == null) {
+    epsilon = tf.test_util.testEpsilon();
+  }
+
+  if (key == 'data') {
```

Comment:
Can we make sure that if key is `data`, the value must be `0|1`?

Reply:
This path is specially for bodypix, which has a special member data in its predict results:
`export type SemanticPersonSegmentation = {
  data: Uint8Array,
  width: number,
  height: number,
  allPoses: Pose[],
};`
The current value for this data member is 0|1. But I am not sure if the same rule applies for the other models.